src/main/io/gps.c

```diff
@@ -960,30 +1419,54 @@ void gpsUpdate(timeUs_t currentTimeUs)
     if (executeTimeUs > gpsStateDurationUs[gpsCurrentState]) {
         gpsStateDurationUs[gpsCurrentState] = executeTimeUs;
     }
+
     schedulerSetNextStateTime(gpsStateDurationUs[gpsData.state]);
 }
 
 static void gpsNewData(uint16_t c)
 {
     if (!gpsNewFrame(c)) {
+        // no new nav solution data
         return;
     }
 
     if (gpsData.state == GPS_STATE_RECEIVING_DATA) {
-        // new data received and parsed, we're in business
-        gpsData.lastLastMessage = gpsData.lastMessage;
-        gpsData.lastMessage = millis();
+        // set the time interval between when we last processed a Nav message.
+        // necessary for NMEA since we do not get a millisecond accurate time stamp with NMEA
+        const uint32_t now = millis();
+        gpsData.navMessageIntervalMs = now - gpsData.lastNavMessage;
+        gpsData.lastNavMessage = now;
+
+#ifdef USE_GPS_UBLOX
+        // Use gpsSol time stamp value from UBX nav solution packets when we get a valid nav solution to process
+        // calculate the interval, handling iTow wraparound at the end of the week
+        const uint32_t weekDurationMs = 604800000; // Number of milliseconds in a week
+        gpsData.gpsNavSolIntervalMs = (gpsSol.time - gpsData.lastNavSolTs + weekDurationMs) % weekDurationMs;
+        gpsData.lastNavSolTs = gpsSol.time;
+        // constrain the interval to 20hz or a value around where we would get a connection failure
+        gpsData.gpsNavSolIntervalMs = constrain(gpsData.gpsNavSolIntervalMs, 50, 2500);
+        DEBUG_SET(DEBUG_GPS_UNIT_CONNECTION, 1, gpsData.gpsNavSolIntervalMs);
+#endif
+
         sensorsSet(SENSOR_GPS);
+        // use the baud rate debug once receiving data
+        DEBUG_SET(DEBUG_GPS_UNIT_CONNECTION, 3, gpsData.navMessageIntervalMs);
     }
-
     GPS_update ^= GPS_DIRECT_TICK;
+    onGpsNewData();
+}
 
-#if DEBUG_UBLOX_INIT
-    debug[3] = GPS_update;
-#endif
+#ifdef USE_GPS_UBLOX
+ubloxVersion_e ubloxParseVersion(const uint32_t version) {
+    for (size_t i = 0; i < ARRAYLEN(ubloxVersion_map); ++i) {
+        if (version == ubloxVersion_map[i].hw) {
+            return (ubloxVersion_e) i;
+        }
+    }
 
-    onGpsNewData();
+    return UBX_VERSION_UNDEF;// (ubloxVersion_e) version;
```

Comment:
(is is possible to differentiate between 'didn't get version' and 'got version, but we don't support it')